### PR TITLE
[functorch] rewrite examples that use make_functional to use functional_call

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91412
* #91282
* __->__ #88851

### CODE REVIEW

**Comment by samdow** [comment]
I can stack below or above the changes to the tests but all of these work locally

File: functorch/examples/dp_cifar10/cifar10_transforms.py [hunk 1]
```diff
@@ -124,12 +120,14 @@ def compute_loss_and_output(weights, image, target):
         # is not to be differentiated. `f'` returns the gradient w.r.t. the loss,
         # the loss, and the auxiliary value.
         grads_loss_output = grad_and_value(compute_loss_and_output, has_aux=True)
+        weights = model.named_parameters()
+        weights = {k: v for k, v in weights}
```
Comment (zou3519) [inline functorch/examples/dp_cifar10/cifar10_transforms.py:124]:
```
weights = {k: v for k, v in model.named_parameters()}
```
is shorter

File: functorch/examples/dp_cifar10/cifar10_transforms.py [hunk 2]
```diff
@@ -124,12 +120,14 @@ def compute_loss_and_output(weights, image, target):
...
         sample_grads, (sample_loss, output) = \
             vmap(grads_loss_output, (None, 0, 0))(weights, images, target)
         loss = sample_loss.mean()
 
-        for grad_sample, weight in zip(sample_grads, model.parameters()):
-            weight.grad_sample = grad_sample.detach()
+        for name, grad_sample in sample_grads.items():
+            weights[name].grad_sample = grad_sample
```
Comment (zou3519) [inline functorch/examples/dp_cifar10/cifar10_transforms.py:130]: where'd the detach() go?

File: functorch/examples/dp_cifar10/cifar10_transforms.py [hunk 3]
```diff
@@ -95,18 +95,14 @@ def train(args, model, train_loader, optimizer, epoch, device):
 
         # Step 1: compute per-sample-grads
 
-        # In order to use functional vmap+grad, we need to be able to
-        # pass the weights to a model.
-        func_model, weights = make_functional(model)
-
```
Comment (zou3519) [inline functorch/examples/dp_cifar10/cifar10_transforms.py:101]: Previously there was a way for users to specify that they don't want the parameters to require_grad when returned from `make_functional`, right? This isn't exhibited in the examples, but it should really be used in the dp-cifar10 example to improve performance. 

I'm not really sure what to do about this

File: functorch/examples/maml_omniglot/maml-omniglot-transforms.py [hunk 4]
```diff
@@ -104,42 +105,38 @@ def main():
...
     for _ in range(n_inner_iter):
         grads = grad(compute_loss)(new_params, buffers, x_spt, y_spt)
-        new_params = [p - g * 1e-1 for p, g, in zip(new_params, grads)]
+        new_params = {k: new_params[k] - g * 1e-1 for k, g, in grads.items()}
 
     # The final set of adapted parameters will induce some
     # final loss and accuracy on the query dataset.
     # These will be used to update the model's meta-parameters.
-    qry_logits = fnet(new_params, buffers, x_qry)
+    qry_logits = functional_call(net, {**new_params, **buffers}, (x_qry,))
```
Comment (zou3519) [inline functorch/examples/maml_omniglot/maml-omniglot-transforms.py:139]: I have some UX nits with the functional_call API:
1. Most users are not going to know that `{**new_params, **buffers}` smashes the dicts together.
2. `functional_call` accepts args and kwargs as a tuple and a dictionary, respectively.

(2) makes functional_call feel a bit bureaucratic. 

We have an opportunity to repurpose functional_call the way we want it. In particular, we are likely going to namespace functional_call into torch.func or whatever we pick. Our version of functional_call can have API improvements to it while we keep around torch.nn.utils.stateless.functional_call for BC.

Here's an idea of what a better functional call could look like:
```
# functional_call(module, dict1, dict2) returns a new callable
functional_call(net, params, buffers)(x_qry)
```

**Review by zou3519 — COMMENTED** [review]
This looks reasonable to me. The API feels more low-level, but it is more faithful to what is going on under the hood. I hope this isn't more difficult to explain to users.

I have many UX nits with the existing functional_call API, so if we agree then we should probably provide a new one that is nicer to use.

**Comment by zou3519** [comment]
cc @Chillee -- curious to hear your opinion on how our examples look after moving from functorch.make_functional to nn.utils.stateless.functional_call.

File: functorch/examples/maml_omniglot/maml-omniglot-transforms.py (hunk 4 above)
Comment (zou3519) [reply functorch/examples/maml_omniglot/maml-omniglot-transforms.py:139]: Do you have a sense of if users have been OK with functional_call accepting args/kwargs as a tuple and dictionary?

File: functorch/examples/dp_cifar10/cifar10_transforms.py (hunk 3 above)
Comment (samdow) [reply functorch/examples/dp_cifar10/cifar10_transforms.py:101]: As it stands the onus is on the user--I'll add this into this example. This should be simple to add to the functorch version of this API (adding on to what you brought up below). Basically this should work if we `.detach()` as we pass these values to vmap or grad, so it should be possible to add a flag to named_parameters or add a helper function (tbh we probably want another helper to avoid the repetition of `{k: v for k, v in model.named_parameters()}`)

File: functorch/examples/maml_omniglot/maml-omniglot-transforms.py (hunk 4 above)
Comment (samdow) [reply functorch/examples/maml_omniglot/maml-omniglot-transforms.py:139]: I did a quick search of issues and Google and didn't see anything pro or against it, but I think one thing to note is that most forward calls just take in a single tensor of inputs. So, I think it's somewhat telling that functional_call now supports just taking in a single tensor. Specifically, I would guess that people kept passing in a single tensor instead of a tuple of a single tensor and got annoyed at either the error message or silent correctness issue

File: functorch/examples/dp_cifar10/cifar10_transforms.py [hunk 5]
```diff
@@ -124,12 +115,14 @@ def compute_loss_and_output(weights, image, target):
         # is not to be differentiated. `f'` returns the gradient w.r.t. the loss,
         # the loss, and the auxiliary value.
         grads_loss_output = grad_and_value(compute_loss_and_output, has_aux=True)
+        weights = {k: v for k, v in model.named_parameters()}
+        detached_weights = {k: v.detach() for k, v in weights.items()}
```
Comment (zou3519) [inline functorch/examples/dp_cifar10/cifar10_transforms.py:119]: Can we add some comments on why we detach the weight here?

**Review by zou3519 — APPROVED** [review]
Looks great!
